shell/browser/notifications/win/windows_toast_activator.cc

```diff
@@ -0,0 +1,414 @@
+#include <vector>
+
+#include <algorithm>
+#include <atomic>
+
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/hash/hash.h"
+#include "base/logging.h"
+#include "base/no_destructor.h"
+#include "base/strings/string_number_conversions_win.h"
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/win/registry.h"
+#include "base/win/scoped_propvariant.h"
+#include "base/win/win_util.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+#include "shell/browser/api/electron_api_notification.h"  // nogncheck - for delegate events
+#include "shell/browser/electron_browser_client.h"
+#include "shell/browser/notifications/notification.h"
+#include "shell/browser/notifications/notification_delegate.h"
+#include "shell/browser/notifications/notification_presenter.h"
+#include "shell/browser/notifications/win/notification_presenter_win.h"
+#include "shell/common/application_info.h"
+
+namespace electron {
+
+namespace {
+
+class NotificationActivatorFactory final : public IClassFactory {
+ public:
+  NotificationActivatorFactory() : ref_count_(1) {}
+  ~NotificationActivatorFactory() = default;
+  IFACEMETHODIMP QueryInterface(REFIID riid, void** ppv) override {
+    if (!ppv)
+      return E_POINTER;
+    if (riid == IID_IUnknown || riid == IID_IClassFactory) {
+      *ppv = static_cast<IClassFactory*>(this);
+    } else {
+      *ppv = nullptr;
+      return E_NOINTERFACE;
+    }
+    AddRef();
+    return S_OK;
+  }
+  IFACEMETHODIMP_(ULONG) AddRef() override { return ++ref_count_; }
+  IFACEMETHODIMP_(ULONG) Release() override {
+    ULONG res = --ref_count_;
+    if (res == 0)
+      delete this;
+    return res;
+  }
+
+  IFACEMETHODIMP CreateInstance(IUnknown* outer,
+                                REFIID riid,
+                                void** ppv) override {
+    if (outer)
+      return CLASS_E_NOAGGREGATION;
+    auto activator = Microsoft::WRL::Make<NotificationActivator>();
+    return activator.CopyTo(riid, ppv);
+  }
+  IFACEMETHODIMP LockServer(BOOL) override { return S_OK; }
+
+ private:
+  std::atomic<ULONG> ref_count_;
+};
+
+NotificationActivatorFactory* g_factory_instance = nullptr;
+bool g_registration_in_progress = false;
+
+std::wstring GetExecutablePath() {
+  wchar_t path[MAX_PATH];
+  DWORD len = ::GetModuleFileNameW(nullptr, path, std::size(path));
+  if (len == 0 || len == std::size(path))
+    return L"";
+  return std::wstring(path, len);
+}
+
+void EnsureCLSIDRegistry() {
+  std::wstring exe = GetExecutablePath();
+  if (exe.empty())
+    return;
+  std::wstring clsid = GetAppToastActivatorCLSID();
+  std::wstring key_path = L"Software\\Classes\\CLSID\\" + clsid;
+  base::win::RegKey key(HKEY_CURRENT_USER, key_path.c_str(), KEY_SET_VALUE);
+  if (!key.Valid())
+    return;
+  key.WriteValue(nullptr, L"Electron Notification Activator");
```

Comment:
No idea, just wondering if we want this to be app-configurable. I wouldn't know if that is user-visible anywhere though.

Reply:
I'm curious too on why we have to write to the registry. I haven't seen that before with the Notification API.